gremlin-go/driver/serializer.go

```diff
@@ -39,23 +41,47 @@ type serializer interface {
 
 // graphBinarySerializer serializes/deserializes message to/from GraphBinary
 type graphBinarySerializer struct {
-	readerClass *graphBinaryReader
-	writerClass *graphBinaryWriter
-	mimeType    string `default:"application/vnd.graphbinary-v1.0"`
+	ser *graphBinaryTypeSerializer
 }
 
 func newGraphBinarySerializer(handler *logHandler) serializer {
-	reader := graphBinaryReader{handler}
-	writer := graphBinaryWriter{handler}
-	return graphBinarySerializer{&reader, &writer, graphBinaryMimeType}
+	serializer := graphBinaryTypeSerializer{NullType, nil, nil, nil, handler}
+	return graphBinarySerializer{&serializer}
 }
 
 const versionByte byte = 0x81
 
+func convertArgs(request *request, gs graphBinarySerializer) (map[string]interface{}, error) {
+	// TODO AN-981: Remote transaction session processor is same as bytecode
+	if request.processor == bytecodeProcessor {
+		// Convert to format:
+		// args["gremlin"]: <serialized args["gremlin"]>
+		gremlin := request.args["gremlin"]
+		switch gremlin.(type) {
+		case bytecode:
+			buffer := bytes.Buffer{}
+			gremlinBuffer, err := gs.ser.write(gremlin, &buffer)
+			if err != nil {
+				return nil, err
+			}
+			request.args["gremlin"] = gremlinBuffer
+			return request.args, nil
+		default:
+			return nil, errors.New(fmt.Sprintf("Failed to find serializer for type '%s'.", reflect.TypeOf(gremlin).Name()))
+		}
+	} else {
+		// Use standard processor, which effectively does nothing.
+		return request.args, nil
+	}
```

Comment:
it looks like everycase in the above `if` has an early return so this else shouldn't be required.

Also, if `bytecodeProcessor` is the "normal" path consider re-writing this as:

```go
if request.processor != bytecodeProcess {
```